pkg/resourcemapping/operator_resource_mapping.go

```diff
@@ -0,0 +1,340 @@
+	if err != nil {
+		return fmt.Errorf("failed to get CR %s for %s in namespace %s: %v", operatorRes, componentKey, resourceNamespace, err)
+	}
+
+	glog.V(2).Infof("Updating CR %s for %s in namespace %s", operatorRes, componentKey, resourceNamespace)
+	updated := false
+	// Update based on each resourceMappingTemplate
+	for _, resourceMappingTemplate := range ormTemplate.resourceMappingTemplates {
+		// Set resourceMappingComponentName to resourceMappingTemplate so as to parse the srcPath and destPath based on text template
+		resourceMappingTemplate[resourceMappingComponentName] = ormTemplate.componentName
+		srcPath, destPath, err := ormClient.parseSrcAndDestPath(resourceMappingTemplate)
+		if err != nil {
+			return fmt.Errorf("failed to update CR %s for %s in namespace %s: %v", operatorRes, componentKey, resourceNamespace, err)
+		}
+
+		// Check if CR needs update from current resourceMappingTemplate. If the values are same under the same srcPath
+		// from origControllerObj and updatedControllerObj, it means resource value is not changed and no need to update
+		// corresponding destPath in CR.
+		newValue, needsUpdate, err := ormClient.needsUpdate(origControllerObj, updatedControllerObj, srcPath)
+		if err != nil {
+			return fmt.Errorf("failed to update CR %s for %s in namespace %s: %v", operatorRes, componentKey, resourceNamespace, err)
+		}
+		if !needsUpdate {
+			glog.V(4).Infof("No changes in controller %s/%s src path '%s'. Skip updating Operator CR %s.",
+				updatedControllerObj.GetKind(), updatedControllerObj.GetName(), srcPath, operatorRes)
+			continue
+		}
+		fields := strings.Split(destPath, ".")
+		if len(fields) < 2 {
+			return fmt.Errorf("failed to update %v to CR %s for %s in namespace %s: '%s' is invalid path", newValue, operatorRes, componentKey, resourceNamespace, destPath)
+		}
+		origCRValue, _, err := util.NestedField(operatorCR, resourceMappingDestPath, destPath)
+		if err != nil {
+			return fmt.Errorf("failed to update %v to CR %s '%s' for %s in namespace %s: %v", newValue, operatorRes, destPath, componentKey, resourceNamespace, err)
+		}
+		err = unstructured.SetNestedField(operatorCR.Object, newValue, fields[1:]...)
+		if err != nil {
+			return fmt.Errorf("failed to update %v to CR %s '%s' for %s in namespace %s: %v", newValue, operatorRes, destPath, componentKey, resourceNamespace, err)
+		}
+		_, err = dynResourceClient.Update(operatorCR, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to update %v to CR %s '%s' for %s in namespace %s: %v", newValue, operatorRes, destPath, componentKey, resourceNamespace, err)
+		}
+		updated = true
+		glog.Infof("Successfully updated CR %s '%s' from %v to %v for %s in namespace %s", operatorRes, destPath, origCRValue, newValue, componentKey, resourceNamespace)
+	}
+	if !updated {
+		return fmt.Errorf("failed to update CR %s for %s in namespace %s: missing resource mapping template", operatorRes, componentKey, resourceNamespace)
```

Comment:
This doesn't seem right. `updated` will remain false if all `resourceMappingTemplates` result in `!needsUpdate`. That seems like a valid case and not an error.

Reply:
Good catch! So when the `operator_resource_mapping.Update` func is triggered, an action is executed from turbo server to make changes on the corresponding resource and there must be some changes (otherwise in `k8s_controller_updater` `update` func, `updated` bool would be true and the func would directly return)

Then if `needsUpdate` here is false, it means there are some changes turbo server wants to make but not defined in the ORM resource mapping templates, for example the fixed field defined in Operator CR but not allowed to be changed. I'm making use of the logic here to throw an error notification and make the action failed because nothing gets changed with this executed action. I'll add a comment here

Ideally, like we discussed before, it'll be helpful if we have a mechanism to avoid market analysis from generated such actions.

Reply:
Its noway anybody looking at this code would know your original intent.. :)
Thanks for adding the comment.. !